doc/specs/#1620 - Installer selector on multiple matches.md

```diff
@@ -0,0 +1,93 @@
+---
+author: Darshan Rander @SirusCodes
+created on: 2022-10-29
+last updated: 2022-10-29
+issue id: 1620
+---
+
+# Spec Title
+
+For [#1620](https://github.com/microsoft/winget-cli/issues/1620)"
+
+## Abstract
+
+The winget.exe client simply lists all the matches and if the user wants to install the app they need to re-run the command with Id.
+
+The UX can be improved here by giving users an option to select the package instead of asking them to re-run the command with Id.
+
+## Inspiration
+
+It is not a good experience to copy Id and then run the command again.
+
+## Solution Design
+
+If we find multiple packages then the user should get a selector.
+
+winget should ask for user input, for example -
+
+```pwsh
+> winget install Powershell
+Multiple packages found matching.
+No. Name       Id                   Source
+-------------------------------------------
+ 1  PowerShell 9MZ1SNWT0N5D         msstore
+ 2  PowerShell Microsoft.PowerShell winget
+
+Please select the packages (eg. 2): <user-input>
+
+Installing <user-input>...
+```
+
+> `<user-input>` - is the input by user
+
+After the user will give the input the installation should start.
+
+## UI/UX Design
+
+The idea is explained in [Solution Design](#solution-design).
```

Comment:
Is tab completion valid for the selector?

Reply:
No

Reply:
Why wouldn't it be? It could very easily cycle through the row indices. I'd rather hit <kbd>Tab</kbd> <kbd>Enter</kbd> than <kbd>1</kbd> <kbd>Enter</kbd> or <kbd>Tab</kbd> <kbd>Tab</kbd> <kbd>Enter</kbd>than <kbd>2</kbd> <kbd>Enter</kbd>

Either way, whether it is supported or not supported should be in the spec

Reply:
I guess it's a personal choice. I prefer `1` `Enter`.
Surely, it can support both, but should it?

What are your opinions @denelon @yao-msft?

Reply:
I'm curious as to whether I would prefer to enter the "number" or the "ID" to disambiguate. My initial thought is that I would type the number.

The number is certainly less typing, but when we think about the tab completion it would likely be less error prone in either case.

If we support both a number or a package identifier, I think I would want the tab to prioritize the package ID first just for the sake of clarity.

In that case, the tab auto complete would be pretty fantastic.

Reply:
Interesting thoughts;  Although that does bring up the edge case where there may be two packages with the same Identifier from different sources, in which case the number would have to be used instead of the package identifier.

Perhaps that calls for some sort of logic like -
> If user enters a row number, treat it as a selection. 
> If the user has entered text, treat it as a new query using only the table results as the source

That would mean that the behavior would be as follows, assuming a table like this:
```pwsh
> winget install Example
Multiple packages found matching.
No. Name            Id              Source
-------------------------------------------
 1  Example         Example.Package custom
 2  Example Beta    Example.Package winget
 3  Example Browser Example.Browser winget

Please enter a row number or refine your query.
> _
```

User enters `2` -
```pwsh
> 2
Found Example Beta [Example.Package] *****
Downloading ******
```
User enters `1` -
```pwsh
> 1
Found Example [Example.Package] *****
Downloading ******
```
User enters `Example.Browser` -
```pwsh
> Example.Browser
Found Example Browser [Example.Browser] *****
Downloading ******
```
User enters `--source custom`
```pwsh
> --source custom
Found Example [Example.Package] *****
Downloading ******
```
User enters `--source winget` -
```pwsh
> --source winget
Multiple packages found matching.
No. Name            Id              Source
-------------------------------------------
 1  Example Beta    Example.Package winget
 2  Example Browser Example.Browser winget

Please enter a row number or refine your query.
> _
```
User enters `Example.Package` -
```pwsh
> Example.Package
Multiple packages found matching.
No. Name            Id              Source
-------------------------------------------
 1  Example         Example.Package custom
 2  Example Beta    Example.Package winget

Please enter a row number or refine your query.
> _
```
User enters `--source msstore` -
```pwsh
> --source msstore
No package found matching selection criteria.
```

Reply:
Thanks @Trenly this looks pretty powerful